### PR TITLE
✨ usePermit hook

### PR DESCRIPTION
**Pull request by insulineru**

### CODE REVIEW

File: src/hooks/usePermit/usePermit.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+    return { ...sig, ...message }
+  }, [provider, token, spender, owner, chainId, deadline])
+
+  const signERC2612Permit = useCallback(async (): Promise<TERC2612PermitMessage & TRSVResponse> => {
+    const message: TERC2612PermitMessage = {
+      owner,
+      spender,
+      value: MAX_UINT256,
+      nonce: await call(provider, token, `${NONCES_FN}${addZeros(24)}${owner.slice(2)}`),
```
Comment (nrjshka) [inline src/hooks/usePermit/usePermit.ts]: I think you can improve this code by creating a new function for this statement

File: src/hooks/usePermit/usePermit.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      holder: owner,
+      spender,
+      nonce: await call(provider, token, `${noncesFn || NONCES_FN}${addZeros(24)}${owner.slice(2)}`),
+      expiry: deadline || MAX_UINT256,
+      allowed: true
+    }
+
+    const domain = await getDomain(provider, token, chainId)
+    const typedData = createTypedDaiData(message, domain)
+    console.log({ typedData, message, domain })
```
Comment (nrjshka) [inline src/hooks/usePermit/usePermit.ts]: Are you sure you wanna log this data?

File: src/hooks/usePermit/usePermit.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    const sig = await signData(provider, owner, typedData)
+
+    return { ...sig, ...message }
+  }, [provider, token, spender, owner, chainId, deadline])
+
+  const signERC2612Permit = useCallback(async (): Promise<TERC2612PermitMessage & TRSVResponse> => {
+    const message: TERC2612PermitMessage = {
+      owner,
+      spender,
+      value: MAX_UINT256,
```
Comment (nrjshka) [inline src/hooks/usePermit/usePermit.ts:32]: You forget to add `deadline` here

Comment (nrjshka) [reply src/hooks/usePermit/usePermit.ts:32]: But I think it's a part of the general problem: I would suggest to separate logic for dai and erc2612 permit to one function and use it twice

File: src/hooks/usePermit/usePermit.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      nonce: await call(provider, token, `${NONCES_FN}${addZeros(24)}${owner.slice(2)}`),
+      deadline: deadline || MAX_UINT256
+    }
+
+    const domain = await getDomain(provider, token, chainId)
+    const typedData = createTypedERC2612Data(message, domain)
+
+    console.log({ typedData, message, domain })
```
Comment (nrjshka) [inline src/hooks/usePermit/usePermit.ts]: Log for dev mode?

File: src/hooks/usePermit/usePermit.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+      // We have a special case for DAI
+      if (isTokenExists(DAI_TOKENS, permitToken)) {
+        return signDaiPermit()
+      } else {
+        return signERC2612Permit()
+      }
+    } else {
+      throw new Error('Token not supported')
+    }
+  }, [token, chainId, signDaiPermit, signERC2612Permit])
```
Comment (nrjshka) [inline src/hooks/usePermit/usePermit.ts]: I would suggest to add something like this:
```
/ const
const AVAILABLE_TOKENS = [‘USDC’, ‘DAI’] as const

…
/ types
type TAvailableTokens = keyof AVAILABLE_TOKENS[number]
type TSupportedTokens = { [key in TAvailableTokens]: Some token type}
…
/ some other consts


const TOKENS: TSupportedTokens = { USDC: USDC_TOKENS, DAI: DAI_TOKENS}
```

So you can use `switch/case` statement here

File: src/hooks/usePermit/usePermit.ts (hunk 2 above)
Comment (insulineru) [reply src/hooks/usePermit/usePermit.ts]: debug, i will delete this after finish work

File: src/hooks/usePermit/usePermit.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+  const getTypedData = useCallback(async () => {
+    switch (getTokenKey(permitToken!)) {
+      case 'DAI':
+        return getDaiPermit()
+      case 'ERC2612':
+        return getERC2612Permit()
+      default:
+        throw new Error('Token not supported')
+    }
```
Comment (nrjshka) [inline src/hooks/usePermit/usePermit.ts:49]: 🔥 🔥 🔥

**Review by nrjshka — APPROVED** [review]

File: src/hooks/usePermit/usePermit.ts (hunk 1 above)
Comment (insulineru) [reply src/hooks/usePermit/usePermit.ts]: Create `getPermitNonce` util

File: src/hooks/usePermit/usePermit.ts (hunk 2 above)
Comment (insulineru) [reply src/hooks/usePermit/usePermit.ts]: Deleted

File: src/hooks/usePermit/usePermit.ts (hunk 4 above)
Comment (insulineru) [reply src/hooks/usePermit/usePermit.ts]: Deleted

File: src/hooks/usePermit/usePermit.ts (hunk 5 above)
Comment (insulineru) [reply src/hooks/usePermit/usePermit.ts]: Refactored

File: src/hooks/usePermit/usePermit.ts (hunk 3 above)
Comment (insulineru) [reply src/hooks/usePermit/usePermit.ts:32]: In DAI we use `expiry`, I added it

File: src/hooks/usePermit/rpc.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
+import type { Web3Provider } from '@ethersproject/providers'
+import type { TRSVResponse } from './types'
+import { splitSignatureToRSV } from './utils'
```
Comment (headfire94) [inline src/hooks/usePermit/rpc.ts:3]: Disadvantage of having "everything in Utils" is that now you have cycle deps between rpc and utils.
"call" imported in utils
This one imported in rpc

Comment (insulineru) [reply src/hooks/usePermit/rpc.ts:3]: I will fixed this circular dependency in next release, thanks 🙏

**Pull request merged**